cmd/ingester/app/consumer/consumer.go

```diff
@@ -97,31 +92,33 @@ func (c *Consumer) Start() {
 
 // Close closes the Consumer and underlying sarama consumer
 func (c *Consumer) Close() error {
-	c.partitionMapLock.Lock()
-	for _, p := range c.partitionIDToState {
-		c.closePartition(p.partitionConsumer)
-		p.wg.Wait()
-	}
-	c.partitionMapLock.Unlock()
-	c.deadlockDetector.close()
+	// Close the internal consumer, which will close each partition consumers' message and error channels.
 	c.logger.Info("Closing parent consumer")
-	return c.internalConsumer.Close()
+	err := c.internalConsumer.Close()
+
+	c.logger.Debug("Closing deadlock detector")
+	c.deadlockDetector.close()
+
+	c.logger.Debug("Waiting for messages and errors to be handled")
+	c.doneWg.Wait()
+
+	return err
 }
 
+// handleMessages handles incoming Kafka messages on a channel
 func (c *Consumer) handleMessages(pc sc.PartitionConsumer) {
 	c.logger.Info("Starting message handler", zap.Int32("partition", pc.Partition()))
 	c.partitionMapLock.Lock()
 	c.partitionsHeld++
 	c.partitionsHeldGauge.Update(c.partitionsHeld)
-	wg := &c.partitionIDToState[pc.Partition()].wg
 	c.partitionMapLock.Unlock()
 	defer func() {
 		c.closePartition(pc)
```

Comment:
Is this `closePartition` still necessary provided `internalConsumer.Close()` closes all partitions?

Reply:
Yup, I think [you're right](https://github.com/bsm/sarama-cluster/blob/v2.1.13/consumer.go#L304), @vprithvi. 👍

However, `closePartition(`) also [increments a metrics counter](https://github.com/jaegertracing/jaeger/blob/404ce68ad033ffaccda063014a60978349f015c8/cmd/ingester/app/consumer/consumer.go#L164), which is used in tests to detect whether if the partition is closed or not; neither the `Consumer` nor `PartitionConsumer` interface appear to offer an "OnClosed" callback or channel to signal successful closure of a partition.

We could instead replace `c.closePartition(pc)` with `c.partitionMetrics(pc.Partition()).closeCounter.Inc(1)` to bump up the metric and remove the `closePartition()` function, if that sounds reasonable to you?